Signal/src/util/OWSScreenLockUI.m

```diff
@@ -307,13 +312,102 @@ - (void)prepareScreenProtectionWithRootWindow:(UIWindow *)rootWindow
     UIWindow *window = [[UIWindow alloc] initWithFrame:rootWindow.bounds];
     window.hidden = YES;
     window.opaque = YES;
-    window.userInteractionEnabled = NO;
     window.windowLevel = CGFLOAT_MAX;
     window.backgroundColor = UIColor.ows_materialBlueColor;
-    window.rootViewController =
-        [[UIStoryboard storyboardWithName:@"Launch Screen" bundle:nil] instantiateInitialViewController];
+
+    UIViewController *viewController = [UIViewController new];
+    viewController.view.backgroundColor = UIColor.ows_materialBlueColor;
+
+    window.rootViewController = viewController;
 
     self.screenBlockingWindow = window;
+    self.screenBlockingViewController = viewController;
+
+    [self updateScreenBlockingWindow:YES shouldHaveScreenLock:NO];
+}
+
+// The "screen blocking" window has three possible states:
+//
+// * "Just a logo".  Used when app is launching and in app switcher.  Must match the "Launch Screen"
+//    storyboard pixel-for-pixel.
+// * "Screen Lock, local auth UI presented". Move the Signal logo so that it is visible.
+// * "Screen Lock, local auth UI not presented". Move the Signal logo so that it is visible,
+//    show "unlock" button.
+- (void)updateScreenBlockingWindow:(BOOL)shouldShowBlockWindow shouldHaveScreenLock:(BOOL)shouldHaveScreenLock
+{
+    OWSAssertIsOnMainThread();
+
+    self.screenBlockingWindow.hidden = !shouldShowBlockWindow;
+
+    UIView *rootView = self.screenBlockingViewController.view;
+    for (UIView *subview in rootView.subviews) {
+        [subview removeFromSuperview];
+    }
+
+    UIImage *image = [UIImage imageNamed:@"logoSignal"];
+    UIImageView *imageView = [UIImageView new];
+    imageView.image = image;
+
+    const CGSize screenSize = UIScreen.mainScreen.bounds.size;
+    const CGFloat shortScreenDimension = MIN(screenSize.width, screenSize.height);
+    const CGFloat imageSize = round(shortScreenDimension / 3.f);
+    [imageView autoSetDimension:ALDimensionWidth toSize:imageSize];
+    [imageView autoSetDimension:ALDimensionHeight toSize:imageSize];
+
+    BOOL shouldShowUnlockButton = (!self.appIsInactive && !self.appIsInBackground && self.didLastUnlockAttemptFail);
+
+    DDLogVerbose(@"%@ updateScreenBlockingWindow. shouldShowBlockWindow: %d, shouldHaveScreenLock: %d, "
+                 @"shouldShowUnlockButton: %d.",
+        self.logTag,
+        shouldShowBlockWindow,
+        shouldHaveScreenLock,
+        shouldShowUnlockButton);
+
+    if (shouldHaveScreenLock) {
+        const CGFloat kVMargin = 50.f;
+        [rootView addSubview:imageView];
+        [imageView autoHCenterInSuperview];
+        [imageView autoPinTopToSuperviewWithMargin:kVMargin];
```

Comment:
So when the screen transitions from the "screen protection" to "lock screen" the logo is moved up higher to make way for some new UI elements.

Some UI feedback on this from the perspective of my iPhone5 (no touch id):

The logo is too high up, it should be closer to the button:
![iphone5](https://user-images.githubusercontent.com/36971200/37993746-94355f88-31dd-11e8-956f-549def16fd0b.png)

The rest of this is TIOLI:

When initially launching the locked app, the "Unlock" button isn't visible and I don't see an alert (no touch id on this device) so it's inexplicable to the user why the logo is not vertically aligned.

Moving the logo top to the top is currently jarring, and it's something that people with this feature enabled will see all the time. Can we animate it?

Also when canceling authentication, I see the system unlock screen without a button for what feels a bit too long. Ideally the button would already be visible when I hit Cancel.

Maybe a complete solution would be:
- Keep the logo vertically centered
- Show the unlock button below center, and render it initially (don't wait for user to cancel, even though the alert will soon be rendered over top of it, it gives the user a hint as to what's going on). That way, if Auth is canceled, unlock button is already there, no need to wait for it to render in
- If and only if an alert is presented, animate up the logo an appropriate amount.